ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1527 @@
-
 	return convd;
 }
 
-
-void
-mbfl_buffer_converter_delete(mbfl_buffer_converter *convd)
+void mbfl_buffer_converter_delete(mbfl_buffer_converter *convd)
 {
-	if (convd != NULL) {
-		if (convd->filter1) {
-			mbfl_convert_filter_delete(convd->filter1);
-		}
-		if (convd->filter2) {
-			mbfl_convert_filter_delete(convd->filter2);
-		}
-		mbfl_memory_device_clear(&convd->device);
-		efree((void*)convd);
+	mbfl_convert_filter_delete(convd->filter1);
+	if (convd->filter2) {
+		mbfl_convert_filter_delete(convd->filter2);
 	}
+	mbfl_memory_device_clear(&convd->device);
+	efree((void*)convd);
 }
 
-int
-mbfl_buffer_converter_illegal_mode(mbfl_buffer_converter *convd, int mode)
+void mbfl_buffer_converter_illegal_mode(mbfl_buffer_converter *convd, int mode)
 {
-	if (convd != NULL) {
-		if (convd->filter2 != NULL) {
-			convd->filter2->illegal_mode = mode;
-		} else if (convd->filter1 != NULL) {
-			convd->filter1->illegal_mode = mode;
-		} else {
-			return 0;
-		}
-	}
-
-	return 1;
+	(convd->filter2 ? convd->filter2 : convd->filter1)->illegal_mode = mode;
 }
 
-int
-mbfl_buffer_converter_illegal_substchar(mbfl_buffer_converter *convd, int substchar)
+void mbfl_buffer_converter_illegal_substchar(mbfl_buffer_converter *convd, int substchar)
 {
-	if (convd != NULL) {
-		if (convd->filter2 != NULL) {
-			convd->filter2->illegal_substchar = substchar;
-		} else if (convd->filter1 != NULL) {
-			convd->filter1->illegal_substchar = substchar;
-		} else {
-			return 0;
-		}
-	}
-
-	return 1;
+	(convd->filter2 ? convd->filter2 : convd->filter1)->illegal_substchar = substchar;
 }
 
 size_t mbfl_buffer_converter_feed(mbfl_buffer_converter *convd, mbfl_string *string)
 {
-	size_t n;
-	unsigned char *p;
-	mbfl_convert_filter *filter;
-	int (*filter_function)(int c, mbfl_convert_filter *filter);
-
-	ZEND_ASSERT(convd);
-	ZEND_ASSERT(string);
-
-	mbfl_memory_device_realloc(&convd->device, convd->device.pos + string->len, string->len/4);
-	/* feed data */
-	n = string->len;
-	p = string->val;
-
-	filter = convd->filter1;
-	if (filter != NULL) {
-		filter_function = filter->filter_function;
-		while (n > 0) {
-			if ((*filter_function)(*p++, filter) < 0) {
-				return p - string->val;
-			}
-			n--;
-		}
-	}
-	return p - string->val;
+	/* Ensure that the buffer has enough space to hold the entire string; this is better
+	 * for performance than repeatedly growing it in small increments */
+	mbfl_memory_device_realloc(&convd->device, convd->device.pos + string->len, string->len / 4);
+	return mbfl_filter_feed_data(string, convd->filter1) - string->val;
 }
 
-
-int
-mbfl_buffer_converter_flush(mbfl_buffer_converter *convd)
+void mbfl_buffer_converter_flush(mbfl_buffer_converter *convd)
 {
-	if (convd == NULL) {
-		return -1;
-	}
-
-	if (convd->filter1 != NULL) {
-		mbfl_convert_filter_flush(convd->filter1);
-	}
-	if (convd->filter2 != NULL) {
-		mbfl_convert_filter_flush(convd->filter2);
-	}
-
-	return 0;
+	mbfl_convert_filter_flush(convd->filter1);
```

Comment:
Why do we only need to flush one of the filters?

Reply:
If you look at the actual code for some flush functions, you will find is that they all check if the next filter in the chain also has a flush function, and call it if so. Then, that one will call the flush function for the next filter in the chain, and so on... so if you flush the first filter in a chain, it's as good as flushing all of them.

If I missed any flush function which does not work in this way, I would love to be told about it. But from what I've seen, that is how it works.

Reply:
Thanks for the explanation, that makes sense!